test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,422 @@
+	for _, machine := range externalMachines {
+		if _, ok := dockerMachineSet[machine.Name()]; !ok {
+			log.V(2).Info("Creating a new DockerMachine for Docker container", "container", machine.Name())
+			dockerMachine, err := r.createDockerMachine(ctx, machine.Name(), cluster, machinePool, dockerMachinePool)
+			if err != nil {
+				return errors.Wrap(err, "failed to create a new docker machine")
+			}
+
+			dockerMachineSet[dockerMachine.Name] = *dockerMachine
+		} else {
+			log.V(4).Info("DockerMachine already exists, nothing to do", "name", machine.Name())
+		}
+	}
+
+	externalMachines, err = docker.ListMachinesByCluster(ctx, cluster, labelFilters)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	externalMachineSet := make(map[string]struct{})
+	for _, externalMachine := range externalMachines {
+		externalMachineSet[externalMachine.Name()] = struct{}{}
+	}
+
+	// Step 2:
+	// Delete any DockerMachines that correspond to a deleted Docker container.
+	// Providers should iterate through the InfraMachines to ensure each one still corresponds to an existing infrastructure instance.
+	// This allows the InfraMachine (and owner Machine) to be deleted and avoid hanging resources when a user deletes an instance out-of-band.
+	for _, dockerMachine := range dockerMachineSet {
+		externalMachine, err := docker.NewMachine(ctx, cluster, dockerMachine.Name, labelFilters)
+		if err != nil {
+			return err
+		}
+		if _, ok := externalMachineSet[externalMachine.Name()]; !ok {
+			log.V(2).Info("Deleting DockerMachine with no underlying infrastructure", "dockerMachine", dockerMachine.Name)
+			if err := r.deleteMachinePoolMachine(ctx, dockerMachine); err != nil {
+				return err
+			}
+
+			delete(dockerMachineSet, dockerMachine.Name)
+		}
+	}
+
+	// Step 3:
+	// This handles the scale down/excess replicas case and the case where a rolling upgrade is needed.
+	// If there are more ready DockerMachines than desired replicas, start to delete the excess DockerMachines such that
+	// - DockerMachines with an outdated Kubernetes version or custom image are deleted first (i.e. the rolling upgrade).
+	// - DockerMachines whose owner Machine contains the clusterv1.DeleteMachineAnnotation are deleted next (to support cluster autoscaler).
+	// Note: we want to ensure that there are always enough ready DockerMachines before deleting anything or scaling down.
+
+	// For each DockerMachine, fetch the owner Machine and copy the clusterv1.DeleteMachineAnnotation to the DockerMachine if it exists before sorting the DockerMachines.
+	// This is done just before sorting to guarantee we have the latest copy of the Machine annotations.
+	readyMachines, err := r.getReadyMachines(ctx, cluster, machinePool, dockerMachinePool)
+	if err != nil {
+		return err
+	}
+	outdatedMachines, err := r.getMachinesWithoutLatestModel(ctx, cluster, machinePool, dockerMachinePool, labelFilters)
```

Comment:
This function is calling getDockerMachines, which we are calling also in L192, L119, and also in a couple of other places. Can we try to read it once, and eventually re-read only when we change the list

Reply:
Adding more on this, I don't think we need two list of machines.
We need only one list with all the machines, we can then call propagate once, order once and then we split the list into two parts before the for loop (without duplicates)